Makefile to check for docker and provide clear error message

Jean-Michel Esnault [9:14 AM]
np, a quick question: how does the payload.tar gets build:  my make distro is failing because this file is not there.
never mind: docker is not running on the box

Mark Sterin [2:39 PM]
Pls file an issue against me to have Makefile checking for docker and provide a.clear error msg

Observed when docker not install/running:
```
[jesnault@s510 km]$ make distro
Creating docker image 'kontain/km:jme-build_km_fedora_from_pipeline'...
make: *** [Makefile:40: distro] Error 1
```

--- a/km/Makefile
+++ b/km/Makefile
@@ -37,6 +37,7 @@ KM_IMAGE_FULL_NAME := kontain/$(EXEC):$(IMAGE_VERSION)
 # Build  a docker image with KM only, to be used as base image if needed
 # We build KM static, put it in a bare docker container, and validate it
 distro: ${BLDEXEC}
+	@$(TOP)make/check-docker.sh
 	@echo "Creating docker image '$(KM_IMAGE_FULL_NAME)'..." ; docker rmi -f $(KM_IMAGE_FULL_NAME) 2>/dev/null
 	tar -C $(dir $(BLDEXEC)) -c $(notdir $(BLDEXEC)) | docker import -m '$(KM_IMAGE_MESSAGE)' -c '$(KM_IMAGE_LABEL)' - $(KM_IMAGE_FULL_NAME)
 	@# Quick validation of the image; it will fail if load_test.km fails

--- a/make/check-docker.sh
+++ b/make/check-docker.sh
@@ -1,0 +1,32 @@
+#!/bin/bash
+#
+# Copyright © 2019 Kontain Inc. All rights reserved.
+#
+# Kontain Inc CONFIDENTIAL
+#
+#  This file includes unpublished proprietary source code of Kontain Inc. The
+#  copyright notice above does not evidence any actual or intended publication of
+#  such source code. Disclosure of this source code or any related proprietary
+#  information is strictly prohibited without the express written permission of
+#  Kontain Inc.
+#
+# Check for docker availability, report issues and exit(1) if something is wrong
+
+RED="\033[31m"
+NOCOLOR="\033[0m"
+
+ver=`docker version -f '{{.Server.Version}}'`
+if [ $? -ne 0 ] ; then
+   echo -e "${RED}Docker is not installed or not running${NOCOLOR}"
+   exit 1
+fi
+
+# Server.Version may be 'dev' (in Moby) or 'NN.xx' in docker distros.
+# But the client version is always NN.xx - so now that we know server is up, we check the client version
+ver=`docker version -f '{{.Client.Version}}'`
+major=`echo $ver| sed 's/\..*$//'`
+echo Docker version $ver major: $major
+if [ -z "$ver" -o $major -lt 18 ] ; then
+      echo -e "${RED}Docker version '$ver' is not supported.  We expect 18 or later${NOCOLOR}"
+      exit 1
+fi;

--- a/make/distro.mk
+++ b/make/distro.mk
@@ -42,10 +42,7 @@ DOCKERFILE_CONTENT := \
 	ENTRYPOINT [ \"/km\"]
 
 distro: all
-	@#check Docker availability
-	@if ! docker ps -aq  > /dev/null ; then \
-		echo -e "${RED}Failed to run Docker - check install and security${NOCOLOR}"; false ; \
-	fi;
+	@$(TOP)make/check-docker.sh
 	@tar -cf $(TAR_FILE) $(PAYLOAD_FILES)
 	@-echo "Cleaning old image"; \
 		docker rmi -f ${PAYLOAD_BRANCH} ${PAYLOAD_LATEST} 2>/dev/null

